monai/apps/detection/networks/utils/anchor_utils.py

```diff
@@ -0,0 +1,326 @@
+    image sizes, inherited from :py:class:`~monai.apps.detection.networks.utils.anchor_utils.AnchorGenerator`
+
+    The module support computing anchors at multiple base anchor shapes
+    per feature map.
+
+    ``feature_map_scales`` should have the same number of elements with the number of feature maps.
+
+    base_anchor_shapes can have an arbitrary number of elements.
+    For 2D images, each element represents anchor width and height [w,h].
+    For 2D images, each element represents anchor width, height, and depth [w,h,d].
+
+    AnchorGenerator will output a set of ``len(base_anchor_shapes)`` anchors
+    per spatial location for feature map ``i``.
+
+    Args:
+        feature_map_scales: scale of anchors for each feature map, i.e., each output level of
+            the feature pyramid network (FPN). ``len(feature_map_scales)`` is the number of feature maps.
+            ``scale[i]*base_anchor_shapes`` represents the anchor shapes for feature map ``i``.
+        base_anchor_shapes: a sequence which represents several anchor shapes for one feature map.
+                For N-D images, it is a Sequence of N value Sequence.
+    """
+
+    __annotations__ = {"cell_anchors": List[torch.Tensor]}
+
+    def __init__(
+        self,
+        feature_map_scales: Union[Sequence[int], Sequence[float]] = (1, 2, 4, 8),
+        base_anchor_shapes: Union[Sequence[Sequence[int]], Sequence[Sequence[float]]] = (
+            (32, 32, 32),
+            (48, 20, 20),
+            (20, 48, 20),
+            (20, 20, 48),
+        ),
+    ) -> None:
+        nn.Module.__init__(self)
+
+        spatial_dims = len(base_anchor_shapes[0])
+        spatial_dims = look_up_option(spatial_dims, [2, 3])
+        self.spatial_dims = spatial_dims
+
+        base_anchor_shapes_t = torch.Tensor(base_anchor_shapes)
+        self.cell_anchors = [self.generate_anchors_using_shape(s * base_anchor_shapes_t) for s in feature_map_scales]
+
+    def generate_anchors_using_shape(
+        self, anchor_shapes: torch.Tensor, dtype: torch.dtype = torch.float32, device: Union[torch.device, None] = None
+    ) -> torch.Tensor:
+        """
+        Compute cell anchor shapes at multiple sizes and aspect ratios for the current feature map.
+
+        Args:
+            anchor_shapes: [w, h] or [w, h, d], sized (N, spatial_dims),
+                represents N anchor shapes for the current feature map.
+            dtype: target data type of the output Tensor.
+            device: target device to put the output Tensor data.
+
+        Returns:
+            For 2D images, returns [-w/2, -h/2, w/2, h/2];
+            For 3D images, returns [-w/2, -h/2, -d/2, w/2, h/2, d/2]
+        """
+        if device is None:
+            device = torch.device("cpu")
+        half_anchor_shapes = anchor_shapes / 2.0
+        base_anchors = torch.cat([-half_anchor_shapes, half_anchor_shapes], dim=1)
+        return base_anchors.round().to(dtype).to(device)
```

Comment:
I came across this generator in detectron, looks like the non-integeters are handled differently https://github.com/facebookresearch/Detectron/blob/1809dd41c1ffc881c0d6b1c16ea38d08894f8b6d/detectron/modeling/generate_anchors.py#L113

Reply:
Yes, they round anchor to int. I don't fully understand the rationale behind it. But we can also round it

Reply:
> I might see your confusion. Detectron uses `TO_REMOVE=1`, So they have a lot of `+1` or `-1` in the code. Is that your concern? Or is `return base_anchors.round().to(dtype).to(device)` your concern? We could also choose to not round it

thanks, I'm not familiar with the details, just wanted to mention this difference, please feel free to ignore this one. 

(but there are a few autodetected issues please fix https://deepsource.io/gh/Project-MONAI/MONAI/run/0660e567-93c6-42b0-b965-16e5500c4937/python/)